src/hardware/ftdi-la/api.c

```diff
@@ -39,89 +40,127 @@ static const uint32_t devopts[] = {
 };
 
-static const uint64_t samplerates[] = {
-	SR_HZ(3600),
-	SR_MHZ(10),
-	SR_HZ(1),
-};
-
 static const struct ftdi_chip_desc ft2232h_desc = {
 	.vendor = 0x0403,
 	.product = 0x6010,
-	.samplerate_div = 20,
+
+	.multi_iface = TRUE,
+	.num_ifaces = 2,
+
+	.base_clock = 120000000u,
+	.bitbang_divisor = 2u,
+	/* My testing on two separate FT2232H chips indicates that channel A
+	 * can run successfully at 15MHz but that channel B will run at 7.5MHz
+	 * if you ask for 15. It's strange, but I'm not gonna turn down three
+	 * extra MHz by limiting both to 12 :) */
+	.max_sample_rates = {15000000u, 12000000u},
+
 	.channel_names = {
-		"ADBUS0", "ADBUS1", "ADBUS2", "ADBUS3",
-		"ADBUS4", "ADBUS5", "ADBUS6", "ADBUS7",
-		/* TODO: BDBUS[0..7] channels. */
-		NULL
+		"ADBUS0", "ADBUS1", "ADBUS2", "ADBUS3", "ADBUS4", "ADBUS5", "ADBUS6", "ADBUS7",
+		"BDBUS0", "BDBUS1", "BDBUS2", "BDBUS3", "BDBUS4", "BDBUS5", "BDBUS6", "BDBUS7",
 	}
 };
 
 static const struct ftdi_chip_desc ft2232h_tumpa_desc = {
 	.vendor = 0x0403,
 	.product = 0x8a98,
-	.samplerate_div = 20,
+
+	.multi_iface = TRUE,
+	.num_ifaces = 1, /* Second interface reserved for UART */
+
+	.base_clock = 120000000u,
+	.bitbang_divisor = 2u,
+	.max_sample_rates = {15000000u, 12000000u},
+
 	/* 20 pin JTAG header */
 	.channel_names = {
 		"TCK", "TDI", "TDO", "TMS", "RST", "nTRST", "DBGRQ", "RTCK",
-		NULL
 	}
 };
 
 static const struct ftdi_chip_desc ft4232h_desc = {
 	.vendor = 0x0403,
 	.product = 0x6011,
-	.samplerate_div = 20,
-	.channel_names = {
-		"ADBUS0", "ADBUS1", "ADBUS2", "ADBUS3",	"ADBUS4", "ADBUS5", "ADBUS6", "ADBUS7",
-		/* TODO: BDBUS[0..7], CDBUS[0..7], DDBUS[0..7] channels. */
-		NULL
-	}
-};
 
-static const struct ftdi_chip_desc ft232r_desc = {
-	.vendor = 0x0403,
-	.product = 0x6001,
-	.samplerate_div = 30,
+	.multi_iface = TRUE,
+	.num_ifaces = 4,
+
+	.base_clock = 120000000u,
+	.bitbang_divisor = 2u,
+	/* TODO: It's likely that channel A (and maybe C or D too) can run at
+	 * 15MHz on the FT4232H just like on the FT2232H, as the two chips use
+	 * the same die internally. However, since I don't have a FT4232 to
+	 * test with, I'm playing it safe and capping them all to 12MHz for
+	 * now. */
+	.max_sample_rates = {12000000u, 12000000u, 12000000u, 12000000u},
+
 	.channel_names = {
-		"TXD", "RXD", "RTS#", "CTS#", "DTR#", "DSR#", "DCD#", "RI#",
-		NULL
+		"ADBUS0", "ADBUS1", "ADBUS2", "ADBUS3",	"ADBUS4", "ADBUS5", "ADBUS6", "ADBUS7",
+		"BDBUS0", "BDBUS1", "BDBUS2", "BDBUS3", "BDBUS4", "BDBUS5", "BDBUS6", "BDBUS7",
+		"CDBUS0", "CDBUS1", "CDBUS2", "CDBUS3", "CDBUS4", "CDBUS5", "CDBUS6", "CDBUS7",
+		"DDBUS0", "DDBUS1", "DDBUS2", "DDBUS3", "DDBUS4", "DDBUS5", "DDBUS6", "DDBUS7",
 	}
 };
 
 static const struct ftdi_chip_desc ft232h_desc = {
 	.vendor = 0x0403,
 	.product = 0x6014,
-	.samplerate_div = 30,
+
+	.multi_iface = TRUE,
+	.num_ifaces = 1,
+
+	/* TODO: This needs validation, since the previous values were weird. */
+	.base_clock = 120000000u,
```

Comment:
Hello,

as this needed validation I have checked an FTDI C232HM-DDHSL-0 devices (which uses a FT232H). I found that the actual sampling rate is about 66 % lower than what is configured in PulseView i.e. an impulse that has duration of 1s (measured on a hardware oscilloscope) is only showing as 666 ms long in PulseView. I have used sample rates of 1 MHz and 100 kHz for this test. I hope this helps.

Best regards,
Otto

Reply:
Thank you so much for the information! I assume those numbers are with these commits applied? Are the timings accurate using the current version of the driver?

I'll try to obtain an FT232H myself so I can dig into what's going on.

Reply:
Thank you for your reply Tom!

I have initially tested the above results on the current (download on 11th June) "[Nightly builds](https://sigrok.org/wiki/Downloads)" of PulseView. I have now tested the same using the current (download on 12th June)  "[Nightly builds](https://sigrok.org/wiki/Downloads)" version of "sigrok-cli". I conducted the test using the 64-Bit Windows versions of PulseView and sigrok-cli.

Please do let me know should you require additional info.

Best regards

Reply:
The nightly builds actually don't include this change, since it hasn't been accepted yet. But that's good news, since the error you're seeing is exactly the amount of error I'd expect from the "weird" divisor value in the current driver that my comment mentions.

Here's a version of [sigrok-cli.exe](https://github.com/sigrokproject/libsigrok/files/8897285/sigrok-cli.zip) that includes this change. Do you mind testing it? It ought to fix the sample rate.

While you're going to the trouble of testing it, would you also mind checking if the 15MHz sample rate works? The 12MHz one definitely should, but the 15MHz one might actually run at 7.5MHz (in which case all your waveforms will appear half as long as they should be). If you validate that it does work, though, I can bump up the maximum sample rate for the FT232H in this pull request :)